### PR TITLE
Implement enforced nonces

### PR DESCRIPTION
**Pull request by AnotherZane**
## Description

Changelog Ref: [Enforced Nonces on Create Message Endpoint](https://discord.com/developers/docs/change-log#enforced-nonces-on-create-message-endpoint)

## Checklist
- [ ] I discussed this PR with the maintainer(s) prior to opening it.
- [x] I read the [contributing guidelines](https://github.com/Quahu/Disqord/blob/master/.github/CONTRIBUTING.md).
- [x] I tested the changes in this PR.

### CODE REVIEW

File: src/Disqord.Core/Entities/Local/Message/LocalMessage.cs [hunk 1]
```diff
@@ -25,6 +25,15 @@ public class LocalMessage : LocalMessageBase, ILocalConstruct<LocalMessage>
     /// </remarks>
     public Optional<string> Nonce { get; set; }
 
+    /// <summary>
+    ///     Gets or sets whether the uniqueness check of a nonce should be enforced.
+    /// </summary>
+    /// <remarks>
+    ///     If set to true and another message with the same nonce was created by the bot in the past few minutes,
+    ///     that message will be returned.
+    /// </remarks>
```
Comment (AnotherZane) [inline src/Disqord.Core/Entities/Local/Message/LocalMessage.cs:34]: Need some feedback on improving these docs

Comment (Quahu) [reply src/Disqord.Core/Entities/Local/Message/LocalMessage.cs:34]: I think I'd just mention that the message won't be sent if another with the same nonce was sent recently.

File: src/Disqord.Core/Entities/Local/Message/Extensions/LocalMessageExtensions.cs [hunk 2]
```diff
@@ -61,15 +61,37 @@ public static TMessage WithReply<TMessage>(this TMessage message, Snowflake mess
...
     /// <returns>
     ///     The input instance.
     /// </returns>
     /// <seealso cref="LocalMessage.Nonce"/>
-    public static TMessage WithNonce<TMessage>(this TMessage message, string nonce)
+    public static TMessage WithNonce<TMessage>(this TMessage message, string nonce, bool? shouldEnforceNonce = null)
         where TMessage : LocalMessage
     {
         message.Nonce = nonce;
+        message.ShouldEnforceNonce = Optional.FromNullable(shouldEnforceNonce);
```
Comment (Quahu) [inline src/Disqord.Core/Entities/Local/Message/Extensions/LocalMessageExtensions.cs:74]:
```suggestion
    public static TMessage WithNonce<TMessage>(this TMessage message, string nonce, bool shouldEnforceNonce = false)
        where TMessage : LocalMessage
    {
        message.Nonce = nonce;
        message.ShouldEnforceNonce = Optional.Conditional(shouldEnforceNonce, true);

```

File: src/Disqord.Core/Entities/Local/Message/Extensions/LocalMessageExtensions.cs [hunk 3]
```diff
@@ -61,15 +61,37 @@ public static TMessage WithReply<TMessage>(this TMessage message, Snowflake mess
...
+    ///     <inheritdoc cref="LocalMessage.ShouldEnforceNonce"/>
+    /// </remarks>
+    /// <param name="message"> The <see cref="LocalMessage"/> instance. </param>
+    /// <param name="shouldEnforceNonce"> Whether the nonce should be enforced. </param>
+    /// <typeparam name="TMessage"> The <see cref="LocalMessage"/> type. </typeparam>
+    /// <returns>
+    ///     The input instance.
+    /// </returns>
+    /// <seealso cref="LocalMessage.ShouldEnforceNonce"/>
+    public static TMessage WithShouldEnforceNonce<TMessage>(this TMessage message, bool shouldEnforceNonce)
```
Comment (Quahu) [inline src/Disqord.Core/Entities/Local/Message/Extensions/LocalMessageExtensions.cs:91]:
```suggestion
    public static TMessage WithShouldEnforceNonce<TMessage>(this TMessage message, bool shouldEnforceNonce = true)

```

File: src/Disqord.Core/Entities/Local/Message/Extensions/LocalMessageExtensions.cs [hunk 4]
```diff
@@ -61,15 +61,37 @@ public static TMessage WithReply<TMessage>(this TMessage message, Snowflake mess
...
+    public static TMessage WithNonce<TMessage>(this TMessage message, string nonce, bool? shouldEnforceNonce = null)
         where TMessage : LocalMessage
     {
         message.Nonce = nonce;
+        message.ShouldEnforceNonce = Optional.FromNullable(shouldEnforceNonce);
+        return message;
+    }
+
+    /// <summary>
+    ///     Sets whether the nonce the uniqueness check of a nonce should be enforced on this message.
```
Comment (Quahu) [inline src/Disqord.Core/Entities/Local/Message/Extensions/LocalMessageExtensions.cs:79]:
```suggestion
    ///     Sets whether the uniqueness check of a nonce should be enforced on this message.

```
typo

**Review by Quahu — APPROVED** [review]

**Review by Quahu — APPROVED** [review]

**Pull request merged**